supervisor/utils/systemd_journal.py

```diff
@@ -0,0 +1,156 @@
+        else:
+            # Other journal fields are serialized in a special binary safe way:
+            # field name, followed by newline,
+            field_name = line.decode("utf-8")
+            data = next(lines)
+            # followed by a binary 64-bit little endian size value,
+            length = struct.unpack_from("<Q", data)[0]
+            # followed by the binary field data,
+            # followed by a newline as separator to the next field.
+            data = data[8:]
+            # handle newlines in binary data
+            while len(data) < length:
+                data += b"\n" + next(lines)
+            if fields and field_name not in fields:
+                # we needed at least iterate up to the end of the data
+                continue
+            field_value = data.decode("utf-8")
+
+        entries[field_name] = field_value
+
+    return entries
+
+
+def formatter(required_fields: list[str]):
+    """Decorate journal entry formatters with list of required fields.
+
+    Helper decorator that can be used for getting list of required fields for a journal
+    formatter function using function.required_fields function attribute.
+    """
+
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            return func(*args, **kwargs)
+
+        wrapper.required_fields = required_fields
+        return wrapper
+
+    return decorator
+
+
+@formatter(["MESSAGE"])
+def journal_plain_formatter(entries: dict[str, str]) -> str:
+    """Format parsed journal entries as a plain message."""
+    return entries["MESSAGE"]
+
+
+@formatter(
+    [
+        "__REALTIME_TIMESTAMP",
+        "_HOSTNAME",
+        "SYSLOG_IDENTIFIER",
+        "_PID",
+        "MESSAGE",
+    ]
+)
+def journal_verbose_formatter(entries: dict[str, str]) -> str:
+    """Format parsed journal entries to a journalctl-like format."""
+    ts = datetime.fromtimestamp(int(entries["__REALTIME_TIMESTAMP"]) / 1e6).isoformat(
+        sep=" ", timespec="milliseconds"
+    )
+
+    identifier = (
+        f"{entries["SYSLOG_IDENTIFIER"]}[{entries["_PID"]}]"
+        if "_PID" in entries
+        else entries["SYSLOG_IDENTIFIER"]
+    )
+
+    return f"{ts} {entries["_HOSTNAME"]} {identifier}: " f"{entries["MESSAGE"]}"
+
+
+async def journal_logs_reader(
+    journal_logs: ClientResponse,
+    log_formatter: LogFormatter = LogFormatter.PLAIN,
+) -> AsyncGenerator[str, None]:
+    """Read logs from systemd journal line by line, formatted using the given formatter."""
+    match log_formatter:
+        case LogFormatter.PLAIN:
+            formatter_ = journal_plain_formatter
+        case LogFormatter.VERBOSE:
+            formatter_ = journal_verbose_formatter
+        case _:
+            raise ValueError(f"Unknown log format: {log_formatter}")
+
+    async with journal_logs as resp:
+        incomplete = b""
+        while not resp.content.at_eof():
+            entry = await resp.content.readuntil(b"\n\n")
```

Comment:
Uh I see, this is tripping in binary data, splitting up messages which belong together :thinking: 

It is a bit weird that you need to go "back and forth" logically to complete the message. Either side (this code nor `parse_journal_export_entry`) has all the information to reliably parse a complete log entry.

Looking at https://systemd.io/JOURNAL_EXPORT_FORMATS/, I think I would try parse this slightly differently:

```python
entries = []
entry = {}
stream: aiohttp.StreamReader = resp.content

while not stream.at_eof():
    entry = await stream.readuntil(b"\n")
    line = entry.decode("utf-8")
    # If empty line, means end of message:
    if not line:
        entries.append(entry)
        entry = {}

    idx = line.find("=")
    if idx == -1:
        # Binary message follows
        length_raw = stream.read(8)
        length = int.from_bytes(length_raw, byteorder='little')
        name = line
        data = stream.read(length)
        if stream.read(1) != b"\n":
	    raise NotAtTheEndOfBinaryData
    else:
        # Regular message
        name = line[:idx]
        data = line[idx+1:]

    entries[name] = data
```

It separates things a bit less, so that is the downside. But you won't need to raise exceptions to complete messages etc.

Reply:
Btw, I went `.decode("utf-8")` here compared to your approach. I am guessing `b"="` might be faster, but I am not sure if there are UTF-8 characters which have a binary `b"="` in it (which are not the equal sign itself). The code relies on reliably detect the equals sign, so I went rather safe and strictly by spec: Check for the separator `=` after decoding UTF-8 ...

Reply:
> I think I would try parse this slightly differently

Nice, thank you! I had to do some adjustments of the pseudocode to the correct form but it seems to work great! I started with a dumb implementation I had to adjust on the way as I added more test cases and haven't gotten back to optimize the core function. This should be definitely more efficient and is quite straightforward in the end too!

> It separates things a bit less, so that is the downside. But you won't need to raise exceptions to complete messages etc.

Yeah, and rewrite the tests quite a bit. But thanks to that, I discovered a better way how to mock the `StreamResponse` API, so all good!

>  I am not sure if there are UTF-8 characters which have a binary b"=" in it

As we only discussed it privately, adding it here for the public record. Should not be an issue, so I'll use bytes instead and decode to UTF only when adding to dict. Citing you:
> Right it seems that Byte 2 through 4 always start with the first bit set, so should never conflict with the equal sign.
> https://en.wikipedia.org/wiki/UTF-8#Encoding

Reply:
Putting it here again for public record - after rewriting the algorithm to one suggested here the performance seems a bit worse (per your testing - I will also do some more tests myself) but still fast enough. Likely the overhead of the frequent reads from the stream is bigger than going back and forth occasionally.   

I added some more fixes, rewrote the the tests to better mock the gateway API and created two builds on DockerHub: `sairon/amd64-hassio-supervisor:logs-v1` with my original algo, and `sairon/amd64-hassio-supervisor:logs-v2` with yours (both also exist as aarch64 image). Everything should be ready for a final review now, we just need to decide between the faster or more straightforward parsing method :)

Reply:
Given the performance difference isn't that drastic (we went from 543kbytes/s -> ~450kbytes/s) it doesn't seem worth to me to take the additional complexity for the gain. The more straightforward path is probably easier to further performance optimize, so I'd go with that as a start.

Probably making the inner most parsing synchronous would fix the speed difference, but it is just PITA as we don't know how much bytes we have to read ahead of time :cry: I tried to find if there is a size limit for a single log entry. If so, we could fill a buffer asynchrounsly to MAX_LOG_ENTRY size, and then parse the individual entries synchronously.

Or implement the thing in Rust :sweat: